examples/greeting/public/plugin.ts

```diff
@@ -27,44 +27,46 @@ export interface GreetingDefinition {
 
 export interface Greeting {
   greetMe: (name: string) => void;
+  label: string;
 }
 
 type GreetingProvider = (def: GreetingDefinition) => Greeting;
 
 const defaultGreetingProvider: GreetingProvider = (def: GreetingDefinition) => ({
   greetMe: (name: string) => alert(`${def.salutation} ${name}${def.punctuation}`),
+  label: def.id,
 });
 
 export interface GreetingStart {
   getGreeting: (id: string) => Greeting;
   getRegisteredGreetings: () => string[];
+  getRegisteredGreetingsAsObjects: () => Greeting[];
 }
 
 export interface GreetingSetup {
   setCustomProvider: (customProvider: GreetingProvider) => void;
-  registerGreeting: (greetingDefinition: GreetingDefinition) => void;
+  registerGreetingDefinition: (greetingDefinition: GreetingDefinition) => () => Greeting;
 }
 
 export class GreetingPlugin implements Plugin<GreetingSetup, GreetingStart> {
-  private greetings: { [key: string]: Greeting } = {};
   private greetingDefinitions: { [key: string]: GreetingDefinition } = {};
   private greetingProvider: GreetingProvider = defaultGreetingProvider;
 
   setup = () => ({
     setCustomProvider: (customProvider: GreetingProvider) =>
       (this.greetingProvider = customProvider),
-    registerGreeting: (greetingDefinition: GreetingDefinition) =>
-      (this.greetingDefinitions[greetingDefinition.id] = greetingDefinition),
+    registerGreetingDefinition: (greetingDefinition: GreetingDefinition) => {
+      this.greetingDefinitions[greetingDefinition.id] = greetingDefinition;
+      return () => this.greetingProvider(greetingDefinition);
+    },
   });
 
   start() {
```

Comment:
I wasn't sure why there was two registries

Reply:
To avoid re-instantiating all `Greeting`s  from `GreetingDefinitions` and preserve state across multiple calls to `getGreeting`.  Might be a problem if the `greetingProvider` wanted to maintain state for each `greetingDefinition`?  Perhaps overkill though, this does look cleaner.